### DISCUSSION TITLE
API: /station/1/requests?searchPhrase=Nirvana / Only works on currently playing playlist?

### DISCUSSION THREAD

**Discussion by BornKillaz** [AzuraCast/AzuraCast#7196] [Q&A] [answered]
Hi.

I'm trying to do a search and retrieve file details, however, I notice that it only returns results from the currently playing playlist.

**/api/station/1/requests?searchPhrase=Nirvana**

Is there any API call that we can use to perform a search on the entire multimedia database? 

Thank you.

**Comment by BornKillaz** [comment]
So, there is no such function, instead we can use our own outside AzuraCast, but it's an heavy one that might kick hard on memory, depending on how much multimedia files you have on your station...

I wish to get a proper search API command in the future, but until then, here is a workaround for anyone who needs it. 

In my case I needed to search for the exact file name, but you can search any of the available fields.

...

```
private function searchFile($filename) {

        // Construct the endpoint to fetch all files
        $endpoint = "/station/{$this->stationId}/files";
        $allFiles = [];

        try {

            // Initial call to fetch data
            $response = $this->apiRequest($endpoint);

            // I know my file would be found on the last ones fetched, so, reverse the files array to start search from the end
            $allFiles = array_reverse($response);

            // Filter files based on the search phrase in the path
            foreach ($allFiles as $file) {
   
                if (basename($file['path']) == $filename) {

                    return $file; // Return immediately when a match is found

                }

            }

            return null; // Return null if no matching file is found

        } catch (Exception $e) {

            // Log the error and throw
            error_log("Error searching media files: " . $e->getMessage());
            throw $e;

        }

    }
```

...

**Reply by BusterNeece** [reply]
@BornKillaz You can definitely search using that API endpoint. Did you try passing `searchPhrase=(phrase)` to it in the query parameters?

**Comment by BornKillaz** [comment]
@BusterNeece, yes I did. And it ignores the query parameter, and returns ALL files.

```
// FOR TEST ONLY
public function search($filename) {

    $endpoint = "/station/{$this->stationId}/files?searchPhrase=".urlencode($filename);

    print_r($this->apiRequest($endpoint));

}
```

**Reply by BusterNeece** [reply]
Oh, internally we use `/files/list?searchPhrase=asdf`. I don't believe `/files` has searching or sorting implemented, though it certainly could in the future.

**Reply by BusterNeece** [reply, marked as answer]
@BornKillaz Eh, it's actually so easy to add search/sort support to that endpoint that I went ahead and did it. It's pushing out in the latest Rolling Release.

**Reply by BornKillaz** [reply]
> @BornKillaz Eh, it's actually so easy to add search/sort support to that endpoint that I went ahead and did it. It's pushing out in the latest Rolling Release.

Thank you @BusterNeece, that is so cool, now I can manage the station pre-recoded shows much easily on the DJ's backend. 

So any plans to be able to perform the search in a specified field, like path for example? 

Something like:
/files?searchPhrase=myfile.mp3&searchField=path

Or is this also implemented already with the latest commit?

**Reply by BusterNeece** [reply]
Currently, no, though we are looking into adding support for "advanced search syntax", where you can specify in your search something like `title="blah" artist~="blah"` and it'll search those respective fields. Currently we're searching for a way to parse such queries in a way we can support within the app.

**Reply by BornKillaz** [reply]
Oh, ok. But is the sort (ORDER BY) parameter already implemented in this endpoint? If so, how to perform it?

/files?searchPhrase=myfile.mp3&(sort parameter here?)

**Reply by BusterNeece** [reply]
@BornKillaz Sorting is implemented in the latest RR too, yes. `?sort=title/artist/path/mtime`, `?sortOrder=ASC/DESC`

**Reply by BornKillaz** [reply]
Oh that is so cool!

So that means I'm able to do like this:

/files?searchPhrase=myfile.mp3&sort=path&sortOrder=DESC

**Reply by BusterNeece** [reply]
You should, yes.

**Reply by Moonbase59** [reply]
Cool things to come, appreciate that!

**Reply by RM-FM** [reply]
I still use internal `/requests` on our website as it ensures that no tracks assigned to jingle or inactive playlists are shown. If there are plans for refactoring the search, it would be great if it becomes possible to search not only with "tags" (e.g. artist, title) but also using a full text search - wildcard or RegEx would be useful too. Meaning that each of the words in the search must be present in artist or title field (possibly others; configurable?).

Example:
`search?artist=Electric Light Orchestra&title=Roll Over Beethoven`
`search?fulltext=Electric Orchestra Beethoven`

**Comment by RM-FM** [comment]
@BusterNeece Great to see that `uploaded_at` is finally available as native file metadata. Any chance to add this to the sort `sortQueryBuilder` for files and requests API endpoints?

**Reply by BillDewey** [reply]
Any chance this will be added to the Search Box on the Public Page as well as the Media Search pages?

**Reply by RM-FM** [reply]
@BillDewey Please vote here: https://features.azuracast.com/suggestions/568040/sort-by-recency-enable-sorting-by-uploaded_at-for-files-and-requests
